docs/src/userguide/wrapping_CPlusPlus.rst

```diff
@@ -683,6 +574,73 @@ An additional C++11-only RTTI-related class, ``std::type_index``, is available
 in ``libcpp.typeindex``.
 
 
+Specify C++ language in setup.py
+================================
+
+Instead of specifying the language and the sources in the source files, it is
+possible to declare them in the :file:`setup.py` file::
+
+   from distutils.core import setup
+   from Cython.Build import cythonize
+
+   setup(ext_modules = cythonize(
+              "rect.pyx",                 # our Cython source
+              sources=["Rectangle.cpp"],  # additional source file(s)
+              language="c++",             # generate C++ code
+         ))
+
+Cython will generate and compile the :file:`rect.cpp` file (from
+:file:`rect.pyx`), then it will compile :file:`Rectangle.cpp`
+(implementation of the ``Rectangle`` class) and link both objects files
+together into :file:`rect.so`, which you can then import in Python using
+``import rect`` (if you forget to link the :file:`Rectangle.o`, you will
+get missing symbols while importing the library in Python).
+
+Note that the ``language`` option has no effect on user provided Extension
+objects that are passed into ``cythonize()``.  It is only used for modules
+found by file name (as in the example above).
+
+The ``cythonize()`` function in Cython versions up to 0.21 does not
+recognize the ``language`` option and it needs to be specified as an
+option to an :class:`Extension` that describes your extension and that
+is then handled by ``cythonize()`` as follows::
+
+   from distutils.core import setup, Extension
+   from Cython.Build import cythonize
+
+   setup(ext_modules = cythonize(Extension(
+              "rect",                                # the extension name
+              sources=["rect.pyx", "Rectangle.cpp"], # the Cython source and
+                                                     # additional C++ source files
+              language="c++",                        # generate and compile C++ code
+         )))
+
+The options can also be passed directly from the source file, which is
+often preferable (and overrides any global option).  Starting with
+version 0.17, Cython also allows passing external source files into the
+``cythonize()`` command this way.  Here is a simplified setup.py file::
+
+   from distutils.core import setup
+   from Cython.Build import cythonize
+
+   setup(
+       name = "rectangleapp",
+       ext_modules = cythonize('*.pyx'),
+   )
+
+And in the .pyx source file, write this into the first comment block, before
+any source code, to compile it in C++ mode and link it statically against the
+:file:`Rectangle.cpp` code file::
+
+   # distutils: language = c++
+   # distutils: sources = Rectangle.cpp
```

Comment:
I think it's worth repeating here that the file path is relative to the working directory of the distutils run (which is usually the project root where the `setup.py` resides).

Reply:
I added a note below.